adaptable logging/measurementfrequency

FransTheekrans opened the pull request:
Hey!

I have your smartpi logger here at home (great thing btw), and have been trying to set it up so it would measure and log at a much higher frequency . Couldn't find anything in the config files, so I started digging into the code and adapted it to suit my needs.
Main changes are in the readout/main.go. I differentiated between loggingrate and samplerate. with samplerate one can define the number of measurements per second, with loggingrate one can define the number of samples that are averaged before the average is written to the database. for example, if you want to measure every second, and log every minute, the samplerate should be 1, and the loggingrate 60.
main changes are in the for loop of the function "pollSmartPi", as you will see.

to do this I had to define an additional parameter in the config file, so this was adapted, together with the config.go-file of the smartpi module to ensure this parameter is parsed.

I'm an engineer, with limited experience in programming, and no experience in golang, so I probably didn't do everything by the book, please excuse me for this. For example, I replaced all references to your github zith references to my fork. this offcourse shouldn't be copied in your main branch. There is probably a better zay, but I don't know that way yet.
thanks and best regards!

Joris

SuperQ commented:
IMO, I would just skip trying to make the database better, and move your data to an external TSDB like Prometheus. This is what I do, and can vary my sample rate from 1s to whatever I feel like.

FransTheekrans commented:
Initially I tried to perform two measurements per second and then average and log every second (I have a three phase measurement setup). This caused quite some gaps in the measurements, so i reverted to one second measuring and logging at the same time. The gaps were still there. CPU load was low, disk and memory usage was good as well, so I suspect I was hitting the limits of the I2C bus as well.
Next I removed my current clamp from N, removed a bunch of unused packages from raspbian, and used a faster SD card. Now I have no more gaps in the measurement. So I'm not convinced it was the I2C bus. (FYI: voltage measurements are not connected on my setup)

I didn't change anything from the raspbian distribution, so I'm using influxDB. If I'm correct, that is also a TSDB (not an IT professional, as I mentioned :D). Why would prometheus be preferred?

Looking at the code, the actions wrt prometheus are done every time a measurement is done. the actions with regards to influxDB are done every 60 seconds in the default master branch, or as defined in loggingrate, in my branch. It would seem more logical to have the same behavior for prometheus and influx.

SuperQ commented:
Both Prometheus and InfluxDB are good TSDBs. Prometheus is just a personal preference for me. I use it for a number of different things (personal and professional).

The difference with Prometheus and InfluxDB is that in Prometheus, the sampling interval is controlled on the server side, not in the client code. This is a big downside to the InfluxDB model.

FransTheekrans commented:
Interesting, thanks for the clarification. This means the loggingrate I defined only affects the grafanaDB. The samplerate affects the Prometheus and Grafana, but on Prometheus should behave as before.

SuperQ commented:
Having adjustable timers is not a bad idea. My personal preference is to sample the data off chip as fast as possible. Then deal with samples to whatever database storage as a separate routine.

One possible way we could improve things is to try and convert some of the measurement gauges to counters. This would allow more accuracy when storing samples at lower intervals. This is typically how we do things in application monitoring with Prometheus.

FransTheekrans commented:
completely agree that would be the ideal approach. If there is enough development capacity on this project, I'd definitely prefer your proposal. If not, this can be a low impact, quick win.
